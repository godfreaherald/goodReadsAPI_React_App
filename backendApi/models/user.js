const mongoose = require("mongoose");
// import mongoose from "mongoose"; to be used in es6 with babel
const bcrypt = require("bcrypt");
const jwt = require("jsonwebtoken");
const uniqueValidator = require("mongoose-unique-validator");
//const secret = process.env.SECRET_KEY;
const secret = "secret";
console.log(secret);
const HOST = process.env.HOST;

const userSchema = new mongoose.Schema(
  {
    email: { type: String, required: true, unique: true },
    isConfirmed: { type: Boolean, default: false },
    confirmationToken: { type: String, default: "" },
    passwordHash: { type: String, required: true },
  },
  { timestamps: true }
);

userSchema.methods.setPassword = function setPassword(password) {
  const salt = bcrypt.genSaltSync(10);

  this.passwordHash = bcrypt.hashSync(password, salt);
};

userSchema.methods.setConfirmationToken = function setConfirmationToken() {
  console.log("test set COnfrimation TOken");
  this.confirmationToken = this.jwtGenerator();
};

userSchema.methods.generateConfirmationURL = function generateConfirmationURL() {
  return `${HOST}/confirmation/${this.confirmationToken}`;
};

userSchema.methods.generateResetPasswordURL = function generateResetPasswordURL() {
  return `${HOST}/resetPassword/${this.hasExpiryJwtGenerator()}`;
};

userSchema.methods.isValidEmail = function isValidEmail(password) {
  return bcrypt.compareSync(password, this.passwordHash);
};

userSchema.methods.jwtGenerator = function jwtGenerator() {
  console.log("test jwt creation during registration");
  return jwt.sign({ email: this.email, isConfirmed: this.isConfirmed }, secret);
};
userSchema.methods.hasExpiryJwtGenerator = function hasExpiryJwtGenerator() {
  console.log("test jwt creation during registration");
  return jwt.sign({ _id: this._id }, secret, { expiresIn: "1h" });
};

userSchema.methods.toJsonAuth = function toJsonAuth() {
  return {
    email: this.email,
    isConfirmed: this.isConfirmed,
    token: this.jwtGenerator(),
  };
};

userSchema.plugin(uniqueValidator, { message: "Email already taken" });

// export default mongoose.model("User", userSchema); to use in es6 with babel

module.exports = mongoose.model("User", userSchema); // es5
