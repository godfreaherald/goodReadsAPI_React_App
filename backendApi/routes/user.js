const router = require("express").Router();
const jwt = require("jsonwebtoken");
const User = require("../models/user");
const mailer = require("../emails/mailer");
const parseError = require("../utils/parseError");
const userController = require("../controllers/userController");

router.post("/signup", userController.signUp);

router.post("/confirm_email", userController.confirmEmail);

router.post("/request_password_reset", userController.requestPasswordReset);

router.post("/reset_password", userController.resetPassword);

module.exports = router;
