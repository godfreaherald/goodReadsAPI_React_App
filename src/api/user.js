import axios from "axios";

//const API_URI = process.env.API_URI;
const api = {
  user: {
    login: credentials => {
      return axios
        .post("http://localhost:8080/user/auth", { credentials })
        .then(res => {
          return res.data.user;
        });
    },
    signup: user => {
      console.log(user);
      return axios
        .post("http://localhost:8080/user/signup", { user })
        .then(res => {
          return res.data.user;
        });
    },
    confirm: token => {
      console.log(token);
      return axios
        .post("http://localhost:8080/user/confirm_email", { token })
        .then(res => {
          return res.data.user;
        });
    },
    requestPasswordReset: email => {
      console.log(email);
      return axios.post("http://localhost:8080/user/request_password_reset", {
        email
      });
    },
    resetPassword: data => {
      console.log(data);
      return axios.post("http://localhost:8080/user/reset_password", {
        data
      });
    }
  }
};

export default api;
