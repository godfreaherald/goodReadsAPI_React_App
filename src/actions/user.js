import { USER_LOGGED_IN, USER_LOGGED_OUT } from "../reduxActionTypes";
import api from "../api/user";

export const loginUser = user => {
  return { type: USER_LOGGED_IN, user };
};

const logoutUser = () => {
  return { type: USER_LOGGED_OUT };
};

export const login = credentials => dispatch => {
  console.log(credentials);
  return api.user.login(credentials).then(user => {
    if (user) {
      localStorage.eccomJwtToken = user.token;
      dispatch(loginUser(user));
    }
  });
};

export const logout = () => dispatch => {
  localStorage.removeItem("eccomJwtToken");
  dispatch(logoutUser());
};
//create user and auto.login.
export const signUp = user => dispatch => {
  console.log(user);
  return api.user.signup(user).then(userRecord => {
    if (userRecord) {
      localStorage.eccomJwtToken = userRecord.token;
      dispatch(loginUser(userRecord));
    }
  });
};

export const confirm = token => dispatch => {
  return api.user.confirm(token).then(userRecord => {
    if (userRecord) {
      localStorage.eccomJwtToken = userRecord.token;
      dispatch(loginUser(userRecord));
    }
  });
};

export const requestPasswordReset = email => () => {
  return api.user.requestPasswordReset(email);
};

export const resetPassword = data => () => {
  return api.user.resetPassword(data);
};
