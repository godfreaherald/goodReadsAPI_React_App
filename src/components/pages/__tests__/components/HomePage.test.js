import React from 'react';
import { shallow,mount } from 'enzyme';
import {Provider} from 'react-redux'
import ConnectedHomePage, {HomePage} from '../../HomePage.js';
import configureStore from 'redux-mock-store'
import { createStore} from "redux";
import renderer from 'react-test-renderer'

import rootReducer from "../../../../reducers/rootReducer";
import { USER_LOGGED_IN, USER_LOGGED_OUT } from "../../../../reduxActionTypes";
import { loginUser, logoutUser } from "../../../../actions/user";


let logoutFn =jest.fn();

 let mockStore =configureStore();


 describe("Snapshot Testing ",()=>{
  it('+++capturing Snapshot of HomePage', () => {
    const renderedValue =  renderer.create(<HomePage isAuthenticated= {true} logout ={logoutFn}/>);
    expect(renderedValue).toMatchSnapshot();
});
})



 
describe('HomePage DUMB component', () => {
  let component;
 beforeEach(()=>{
   component = shallow(<HomePage isAuthenticated= {true} logout ={logoutFn} />);
 });

   it(" Dumb page is rendered" ,()=>{
       expect(component.length).toEqual(1);
   });
   it(" snapshot is matching ",()=>{
     expect(component).toMatchSnapshot();
   });

  it('Has Logout button when authenticated', () => {
    
    const text = component.find('button').text();
   expect(text).toEqual('Logout');
   
  });

  it(' Logout button click should call logout function', () => {
    
     component.find('button').simulate('click');
   expect(logoutFn).toHaveBeenCalled();
   
  });
});


describe('using redux store  to', () => {
  let store ;
  //let wrapper;
  let innitialState = { isAuthenticated:true}
  beforeEach(()=>{
   store = mockStore(innitialState)

it('+++ check action on dispatching ', () => {
  let action
  store.dispatch(loginUser({name:"herald"}))
  store.dispatch(logoutUser())
  action = store.getActions()
  expect(action[0].type).toBe(USER_LOGGED_IN)
  expect(action[1].type).toBe(USER_LOGGED_OUT)
});
});


// describe('ConnectedHomePage component using Shallow', () => {
//   let store ;let component;
//   let innitialState = { isAuthenticated:true}
//   beforeEach(()=>{
//    store = mockStore(innitialState)
//     component = shallow(<ConnectedHomePage store= {store} logout ={logoutFn} />);
//   });
 
//     it(" is rendered properly" ,()=>{
//         expect(component.length).toEqual(1);
//     });
//     it(" has isAuthenticated Prop equal to the initial state ",()=>{
//       expect(component.prop('isAuthenticated')).toEqual(innitialState.isAuthenticated);
//     });

//     it(" snapshot is matching ",()=>{
//       expect(component).toMatchSnapshot();
//     });
 
//    it('Has Logout button when authenticated', () => {
//      if(innitialState.isAuthenticated === true){

//        const text = component.find('button').text();
//       expect(text).toEqual('Logout');
//      }
//      else {
//       const text = component.find('button').text();
//       expect(text).toEqual('Login');
//      }
    
//    });
//  });




 //describe('ConnectedHomePage component using mount + provider', () => {

  // let store ;
  // //let wrapper;
  // let innitialState = { isAuthenticated:true}
  // beforeEach(()=>{
  //  store = mockStore(innitialState)
   //wrapper = mount( <Provider store={store}><ConnectedHomePage logout ={logoutFn} /></Provider>);
  });
 
//     it(" is rendered properly" ,()=>{
//         expect(wrapper.find(ConnectedHomePage).length).toEqual(1);
//     });
//     it(" has isAuthenticated Prop equal to the initial state ",()=>{
//       expect(wrapper.find(HomePage).prop('isAuthenticated')).toEqual(innitialState.isAuthenticated);
//     });

//     it(" snapshot is matching ",()=>{
//       expect(component).toMatchSnapshot();
//     });
 
//    it('Has Logout button when authenticated', () => {
//      if(innitialState.isAuthenticated === true){

//        const text = wrapper.find('button').text();
//       expect(text).toEqual('Logout');
//      }
//      else {
//       const text = component.find('button').text();
//       expect(text).toEqual('Login');
//      }
    
//    });

//    it('+++ check action on dispatching ', () => {
//     let action
//     store.dispatch(loginUser({name:"herald"}))
//     store.dispatch(logoutUser())
//     action = store.getActions()
//     expect(action[0].type).toBe(USER_LOGGED_IN)
//     expect(action[1].type).toBe(USER_LOGGED_OUT)
// });
//  });


//  describe('>>>H O M E --- REACT-REDUX (actual Store + reducers) more of Integration Testing',()=>{
  
//   let innitialState = { isAuthenticated:true}
//   let store,wrapper
//   beforeEach(()=>{
//       store = createStore(rootReducer)
//       wrapper = mount( <Provider store={store}><ConnectedHomePage logout ={logoutFn} /></Provider> )
//   });

//   it('+++ check Prop matches with initialState', () => {
//       store.dispatch(logoutUser())
//      expect(wrapper.find('button').text().toBe('Login')); 
// });

// });

