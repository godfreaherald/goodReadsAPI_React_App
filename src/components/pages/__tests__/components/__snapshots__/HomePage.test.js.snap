// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage DUMB component  snapshot is matching  1`] = `
<div>
  <h1>
     HomePage
  </h1>
  <button
    onClick={[Function]}
  >
    Logout
  </button>
</div>
`;

exports[`Snapshot Testing  +++capturing Snapshot of HomePage 1`] = `
<div>
  <h1>
     HomePage
  </h1>
  <button
    onClick={[Function]}
  >
    Logout
  </button>
</div>
`;
